From 4d412b1d64855de58d8ab681de0c60b5e5061584 Mon Sep 17 00:00:00 2001
From: Michael Stapelberg <stapelberg@debian.org>
Date: Wed, 8 Apr 2026 11:05:59 +0800
Subject: [PATCH 09/12] DEBIAN: Open the PAM session in the parent process.

This is necessary when logind is running. pam_systemd communicates with
logind and sets up a file descriptor that represents the session. When
this file descriptor is closed, logind will close the session and kill
all processes as part of that. The file descriptor is marked CLOEXEC, so
without this patch, the file descriptor gets opened, then wdm exec()s
the session and the file descriptor is closed, leading to logind
immediately killing the session. This patch makes wdm do PAM session
handling the same way that xdm does it.

Link: https://salsa.debian.org/debian/wdm/-/blob/debian/1.28-27/debian/patches/09_pam_session.patch
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 src/wdm/session.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/wdm/session.c b/src/wdm/session.c
index d554ee7..46a612b 100644
--- a/src/wdm/session.c
+++ b/src/wdm/session.c
@@ -550,6 +550,7 @@ StartClient (
 	WDMDebug("\n");
     }
 #ifdef USE_PAM
+    if (pamh) pam_open_session(pamh, 0);
 #endif    
     switch (pid = fork ()) {
     case 0:
@@ -599,8 +600,6 @@ StartClient (
 		for(i = 0; pam_env && pam_env[i]; i++) {
 			verify->userEnviron = WDMPutEnv(verify->userEnviron, pam_env[i]);
 		}}
-
-		pam_open_session(pamh, 0);
 	}
 #endif
 	if (setuid(verify->uid) < 0)
-- 
2.53.0

